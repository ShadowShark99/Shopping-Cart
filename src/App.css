.layout {
  display: grid;
  width: 100vw;
  height: 100vh;
  grid-template-columns: 200px 1fr;
}
