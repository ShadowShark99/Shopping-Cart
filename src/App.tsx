import { useEffect, useState } from 'react'
import './App.css'
import Navigation from './components/Navigation'
import { Outlet } from 'react-router-dom'

function App() {
  
  //fetch('https://fakestoreapi.com/products')
  // .then(response => response.json())
  // .then(data => console.log(data));
  console.log("app loading");

  const newArray = (n: number) => {
    if(n <= 0)
      return [];
    const ret = [];
    for(let i = 0; i < n; i++)
    {
      ret.push(0);
    }
    return ret;
  };

  type Product = {
    id: number;
    title: string;
    price: number;
    description: string;
    category: string;
    image: string;
    rating?: {
      rate: number;
      count: number;
    };
  };

  const [products, setProducts] = useState<Product[]>([]);
  const [bought, setBought] = useState(newArray(20));

  const toggleBuy = (index: number) => {
    console.log("buy");
    if(index < 0 || index >= 20)
      return;
    const newBought = [...bought];
    newBought[index] = 1 - newBought[index];
    setBought(newBought);
  };
  
  useEffect(() => {
    const fetchProducts = async () => {
      //const response = await fetch('https://fakestoreapi.com/products');
      const response = await fetch('https://fakestoreapi.com/products');
      const result = await response.json();
      console.log(result);
      setProducts(result);
    };
    fetchProducts();
  }, []);

  //calc cost of products array
  const calcCost = () => {
    let sum = 0;
    for(let i = 0; i < products.length; i++)
    {
      sum = bought[i] ? sum + products[i].price : sum;
    }
    return sum;
  };

  const items = () => {
    let ret = 0;
    for(let i = 0; i < 20; i++)
    {
      ret += bought[i];
    }
    return ret;
  };

  const numberOfItems = items();
  const totalCost = calcCost();


  return (
    <div className ="layout">
      <Navigation items={numberOfItems} totalCost = {totalCost}></Navigation>
      <Outlet context={{products, toggleBuy}}></Outlet>
    </div>
  )
}

export default App
