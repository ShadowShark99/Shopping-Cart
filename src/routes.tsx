import App from "./App";
import ErrorPage from "./components/ErrorPage";
import Home from "./components/Home";
import Shop from "./components/Shop";

const routes = [
  {
    path: "/",
    element: <App />,
    errorElement: <ErrorPage/>,
    children: [
      {index: true, element: <Home></Home>},
      { path: "shop/:page", 
        element: <Shop></Shop>
      }
    ]
  },
  //  {
  //    path: "shop",
  //    element: <Shop />,
  //  },
];

export default routes;