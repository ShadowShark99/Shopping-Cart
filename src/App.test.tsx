// App.test.jsx

import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "./App";
import { createBrowserRouter, MemoryRouter, RouterProvider } from "react-router-dom";
import routes from "./routes";

describe("App component", () => {
  it("renders the app", () => {
    // since screen does not have the container property, we'll destructure render to obtain a container for this test
     const { container } = render(<MemoryRouter><App /></MemoryRouter>);
     expect(container).toMatchSnapshot();
  });

  it("Shopping page", async () => {
    const router = createBrowserRouter(routes);

    const user = userEvent.setup();

    render(<RouterProvider router={router}></RouterProvider>);
    const button = screen.getByRole("link", { name: "Shop" });

    await user.click(button);

    const heading = await screen.getByRole("heading", {level: 2});
    // for(let i = 0; i < headings.length; i++)
    // {
    //   console.log(i);
    // }
    // expect(screen.getByRole("heading").textContent).toMatch(/Hi, want to buy something?/i);
    expect(heading.textContent).toMatch(/Hi, want to buy something?/i);
  });
});
