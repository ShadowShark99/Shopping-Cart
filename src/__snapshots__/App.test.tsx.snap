// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders the app 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="nav"
    >
      <h3>
        Let's Shop!
      </h3>
      <div
        class="nav-options"
      >
        <a
          class="home-button"
          data-discover="true"
          href="/"
        >
          Home
        </a>
        <div>
          <a
            class="home-button"
            data-discover="true"
            href="/shop"
          >
            Shop
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
